File: nelman-kmcLATTICEMODEL/source/Design.cpp
/**
 * \file   Design.hpp
 * \date   May 2013
 * \author Sanne Abeln
 * \brief  Defines class Design
 */


#include "Design.hpp"
#include "Lattice.hpp"
#include "Chain.hpp"
#include "Residue.hpp"
#include "AA.hpp"
#include "StatsMethods.hpp"
#include "App.hpp"
#include "AADistr.hpp"
#include "Moves.hpp"

#include <fstream>
#include <iomanip>
#include <sstream>


double countAA[MAXAA];

double pChangeStrandCoil2=0.5;

double systemVar=0;

int numRes=0;


Design::Design(){
  mainLattice = new Lattice;
  mainLattice->setAA(App::fn_aa,App::indx_water); 
  //// set AA distribution
  if(App::AAdistribution){
    aaDistr = new AADistr();
    aaDistr->init(App::fn_AAdistribution,mainLattice->aaInt);
  }
}


void Design::init(string fn_in){
  mainLattice->readPDBMultiChain(fn_in);
  if(App::AAdistribution){
    aaDistr->setSequence(mainLattice);
  }else{
    initCountAA();
  }
  // mainLattice->stats.getLatticeStats(mainLattice);
  mainLattice->stats.printCout();

  /// set random spins
  for(int cn =0;cn<mainLattice->nChains;cn++){
    if(! mainLattice->chains[cn]->frozen && App::flipSpins)
      mainLattice->chains[cn]->setRandomSpins();
  }
}

void Design::initOldFormat(string fn_in){
  mainLattice->oldReadPDBMultiChain(fn_in);
  initCountAA();
  


  /// set random spins
  for(int cn =0;cn<mainLattice->nChains;cn++){
    if(! mainLattice->chains[cn]->frozen && App::flipSpins)
      mainLattice->chains[cn]->setRandomSpins();
  }
}



void Design::designProcedure(double beta, long numSteps){

  mainLattice->setBetaMoves(beta);
  cout<< "energy Beta "<<beta<<endl;

  //for(int cn =0;cn<mainLattice->nChains;cn++){
  //  if(! mainLattice->chains[cn]->frozen && App::flipSpins)
  //  mainLattice->chains[cn]->setRandomSpins();
  // }
  mainLattice->stats.getLatticeStats(mainLattice);

  for(long i=0;i<numSteps;i++){
    // change AA
    int cn = (int) floor(mainLattice->nChains * drand48());
    if(!mainLattice->chains[cn]->frozen){ 
      changeAA(cn);
      //if(i < (numSteps/2)){
      if(App::flipSpins){
	if (drand48() < pChangeStrandCoil2){
	  // change state
	  Chain * c= mainLattice->chains[(int) floor(mainLattice->nChains * drand48())];
	  Moves::changeStrandCoil(mainLattice,c);
	}else{ 
	  // change spin
	  Chain * c= mainLattice->chains[(int) floor(mainLattice->nChains * drand48())];
	  Moves::shuffleSpin(mainLattice,c);
	}
      }
    }
  }
  mainLattice->stats.printCout();
  mainLattice->checkStats();
  if(App::AAdistribution){
    aaDistr->check();
  }else{
    checkVariance();
  }
}


void Design::finalize(){
  std::stringstream fn_out_design_out;
  std::stringstream fn_out_design_pdb;
  
  fn_out_design_out << "design" <<myRank <<".out"; 
  fn_out_design_pdb << "design" <<myRank <<".pdb"; 
  
  writeDesignStats(fn_out_design_out.str());
  mainLattice->writePDBMultiChain(fn_out_design_pdb.str());
  
  mainLattice->stats.printCout();
  mainLattice->checkStats();
  if(App::AAdistribution){
    aaDistr->check();
  }else{
    checkVariance();
  }

}


void Design::checkVariance(){
 if(systemVar != getSystemVar()){
    cout <<"ERROR in design process accummulative variance:"<<systemVar;
    cout <<" does not match calc variance:" <<getSystemVar()<<endl;
    //exit(1);
 }
}

void Design::changeAA(int chain){
  Chain * c = mainLattice->chains[chain];
  int n = (int) floor(c->N * drand48());
  Residue * res = c->residues[n];
  int oldAA = res->aa;
  int newAA;
  do{  newAA = (int) floor(AA::NUMAA * drand48());}while(newAA == AA::designWater);
  
 
  //// draw AA acid with respect to correct distribution
  double localVar =0; double dist=0;
  if(App::AAdistribution){
    dist = aaDistr->getDistanceSubst(oldAA,newAA);
    double acc = exp(-dist / App::designT);
    if(!(drand48()<acc)){
      return;
    }
    
  }else{
    localVar = countAA[oldAA]/(countAA[newAA]+1.0);
    double accVar =pow(localVar ,App::designT);
    if(!(drand48()<accVar)){
      return;
    }
  }
  
  Stats oldLocalStats; 
  Stats newLocalStats; 
  oldLocalStats.localStats(res,res->pos,res->spin,res->state,oldAA,mainLattice);
  newLocalStats.localStats(res,res->pos,res->spin,res->state,newAA,mainLattice);
  Stats newLatticeStats =  mainLattice->stats.delta(newLocalStats,oldLocalStats);
  
  int dE= newLocalStats.getDeltaE(oldLocalStats);
  double boltz = exp(((-(float) (dE)) )*  mainLattice->getBetaMoves());
  bool accept=  dE<=0 || (drand48() < boltz);    
  
  if(accept){
    // update AA
    res->aa=newAA;
    // update energy
    mainLattice->stats = newLatticeStats;
    // update AAcount

    if(App::AAdistribution){
      aaDistr->substitute(oldAA,newAA,dist);
    }else{
      countAA[oldAA]--;
      countAA[newAA]++;
      systemVar = systemVar *localVar;
    }
  }
}



void Design::initCountAA(){
  numRes=0;
  for(int i=0;i<AA::NUMAA;i++){
    countAA[i]=0;
  }

  for(int nc=0;nc<mainLattice->nChains;nc++){
    Chain * ch = mainLattice->chains[nc];
    for (int n=0;n<ch->N;n++){
      Residue * res =ch->residues[n];
      countAA[res->aa]++;
      numRes++;
    }
  }
  systemVar = getSystemVar();
}


double factorial(double number) {
  double temp;
  if(number <= 1.0) return 1.0;
  temp = number * factorial(number - 1.0);
  return temp;
}

double Design::getSystemVar(){
  double ni=1.0;

  for(int aa=0;aa<AA::NUMAA;aa++){
    if(countAA[aa]){
      ni *= factorial(countAA[aa]);
    }
  }
  //cout<<usedRes<<endl;

  //return (factorial((double)numRes))/((double) ni);
  return (1.0)/((double) ni);
}


void Design::writeDesignStats(string s){
  ofstream outs;
  cout<< "printing to "<< s<<endl;
  outs.open(s.c_str()); 

  //designT
  outs<<"designT:"<<App::designT<<endl;
  //energyT
  outs<<"energyT:"<<mainLattice->getBetaMoves()<<endl;
  //sequence
  for(int nc=0;nc<mainLattice->nChains;nc++){
    Chain * ch = mainLattice->chains[nc];
    outs<<"chain no:"<< nc<<endl;
    for (int n=0;n<ch->N;n++){
      Residue * res =ch->residues[n];
      outs<<AA::int2aa[res->aa]<<" ";
    }
    outs<<endl;
  }
  //Stats internal energy 
  outs<< mainLattice->stats.print2string();
 
  //sequence variance
  if(App::AAdistribution){
    outs<< aaDistr->toString() <<endl;
  }else{
    outs<<"systemsVar:"<<systemVar<<endl;
    //AA count
    outs<<"AA count:"<< endl;
    for(int aa=0;aa<AA::NUMAA;aa++){
      outs <<AA::int2aa[aa]<<" "<<countAA[aa]<<", ";
    }
    outs<<endl;
  }
 outs.close();
}

File: nelman-kmcLATTICEMODEL/source/Design.py
from typing import List, Dict, Optional
import random
import math
import sys
from AA import AA
from Lattice import Lattice
from AADistr import AADistr
from Stats import Stats
from Moves import changeStrandCoil, shuffleSpin

MAXAA = 21

class Design:
    def __init__(self):
        self.main_lattice = Lattice()
        self.main_lattice.setAA(App.fn_aa, App.indx_water)
        self.aa_distr = None
        self.count_aa = [0] * MAXAA
        self.system_var = 0
        self.num_res = 0
        
        if App.AAdistribution:
            self.aa_distr = AADistr()
            self.aa_distr.init(App.fn_AAdistribution, self.main_lattice.aaInt)
    
    def init(self, fn_in: str):
        self.main_lattice.readPDBMultiChain(fn_in)
        if App.AAdistribution:
            self.aa_distr.setSequence(self.main_lattice)
        else:
            self.initCountAA()
        
        # Set random spins
        for cn in range(self.main_lattice.nChains):
            if not self.main_lattice.chains[cn].frozen and App.flipSpins:
                self.main_lattice.chains[cn].setRandomSpins()
    
    def designProcedure(self, beta: float, num_steps: int):
        self.main_lattice.setBetaMoves(beta)
        print(f"energy Beta {beta}")
        
        for i in range(num_steps):
            # Change AA
            cn = random.randint(0, self.main_lattice.nChains - 1)
            if not self.main_lattice.chains[cn].frozen:
                self.changeAA(cn)
                
                if App.flipSpins and random.random() < 0.5:
                    if random.random() < 0.5:
                        # Change state
                        c = self.main_lattice.chains[random.randint(0, self.main_lattice.nChains - 1)]
                        changeStrandCoil(self.main_lattice, c)
                    else:
                        # Change spin
                        c = self.main_lattice.chains[random.randint(0, self.main_lattice.nChains - 1)]
                        shuffleSpin(self.main_lattice, c)
    
    def changeAA(self, chain: int):
        c = self.main_lattice.chains[chain]
        n = random.randint(0, c.N - 1)
        res = c.residues[n]
        old_aa = res.aa
        new_aa = old_aa
        while new_aa == old_aa or new_aa == AA.designWater:
            new_aa = random.randint(0, AA.NUMAA - 1)
        
        # Check distribution constraints
        local_var = 0
        dist = 0
        if App.AAdistribution:
            dist = self.aa_distr.getDistanceSubst(old_aa, new_aa)
            acc = math.exp(-dist / App.designT)
            if random.random() >= acc:
                return
        else:
            local_var = self.count_aa[old_aa] / (self.count_aa[new_aa] + 1.0)
            acc_var = pow(local_var, App.designT)
            if random.random() >= acc_var:
                return
        
        # Calculate energy changes
        old_stats = Stats()
        new_stats = Stats()
        old_stats.localStats(res, res.pos, res.spin, res.state, old_aa, self.main_lattice)
        new_stats.localStats(res, res.pos, res.spin, res.state, new_aa, self.main_lattice)
        
        dE = new_stats.getDeltaE(old_stats)
        boltz = math.exp(-dE * self.main_lattice.getBetaMoves())
        accept = dE <= 0 or random.random() < boltz
        
        if accept:
            res.aa = new_aa
            self.main_lattice.stats = self.main_lattice.stats.delta(new_stats, old_stats)
            
            if App.AAdistribution:
                self.aa_distr.substitute(old_aa, new_aa, dist)
            else:
                self.count_aa[old_aa] -= 1
                self.count_aa[new_aa] += 1
                self.system_var *= local_var
    
    def initCountAA(self):
        self.num_res = 0
        self.count_aa = [0] * AA.NUMAA
        
        for nc in range(self.main_lattice.nChains):
            ch = self.main_lattice.chains[nc]
            for n in range(ch.N):
                res = ch.residues[n]
                self.count_aa[res.aa] += 1
                self.num_res += 1
        
        self.system_var = self.getSystemVar()
    
    def getSystemVar(self) -> float:
        ni = 1.0
        for aa in range(AA.NUMAA):
            if self.count_aa[aa]:
                ni *= math.factorial(self.count_aa[aa])
        return 1.0 / ni
    
    def writeDesignStats(self, filename: str):
        with open(filename, 'w') as f:
            f.write(f"designT:{App.designT}\n")
            f.write(f"energyT:{self.main_lattice.getBetaMoves()}\n")
            
            # Write sequence
            for nc in range(self.main_lattice.nChains):
                ch = self.main_lattice.chains[nc]
                f.write(f"chain no:{nc}\n")
                for n in range(ch.N):
                    res = ch.residues[n]
                    f.write(f"{AA.int2aa[res.aa]} ")
                f.write("\n")
            
            # Write stats
            f.write(self.main_lattice.stats.print2string())
            
            # Write distribution info
            if App.AAdistribution:
                f.write(self.aa_distr.toString())
            else:
                f.write(f"systemsVar:{self.system_var}\n")
                f.write("AA count:\n")
                for aa in range(AA.NUMAA):
                    f.write(f"{AA.int2aa[aa]} {self.count_aa[aa]}, ")
                f.write("\n")
